examples/gcs-to-bq-serverless-services/gcs2bq-spark/src/main/java/com/example/GCS2BQ.java

```diff
@@ -0,0 +1,167 @@
+import com.google.cloud.pubsub.v1.Publisher;
+import com.google.cloud.spark.bigquery.repackaged.org.json.JSONObject;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.protobuf.ByteString;
+import com.google.pubsub.v1.PubsubMessage;
+import com.google.pubsub.v1.TopicName;
+import java.io.IOException;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SaveMode;
+import org.apache.spark.sql.SparkSession;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import picocli.CommandLine;
+import picocli.CommandLine.Command;
+import picocli.CommandLine.Option;
+import picocli.CommandLine.Spec;
+
+@Command(name = "defaults", mixinStandardHelpOptions = true, version = "defaults 0.1")
+public class GCS2BQ implements Runnable {
+
+  public static final Logger LOGGER = LoggerFactory.getLogger(GCS2BQ.class);
+
+  @Option(
+      names = {"--projectId"},
+      required = true,
+      description = "Project Id to run serverless spark")
+  String projectId;
+
+  @Option(
+      names = {"--inputFileLocation"},
+      required = true,
+      description = "GCS location of the file")
+  String inputFileLocation;
+
+  @Option(
+      names = {"--bqDataset"},
+      required = true,
+      description = "BigQuery Dataset")
+  String bqDataset;
+
+  @Option(
+      names = {"--bqTable"},
+      required = true,
+      description = "BigQuery Table")
+  String bqTable;
+
+  @Option(
+      names = {"--bqTempBucket"},
+      required = true,
+      description = "BigQuery Temp Bucket")
+  String bqTempBucket;
+
+  @Option(
+      names = {"--deadLetterQueue"},
+      required = true,
+      description = "Dead letter Topic")
+  String dlq;
+
+  @Spec CommandLine.Model.CommandSpec spec;
+
+  @Override
+  public void run() {
+    SparkSession spark = null;
+    final String inputFileFormat = "CSV";
+    try {
+      spark = SparkSession.builder().appName("GCS to Bigquery load").getOrCreate();
+      Dataset<Row> inputData =
+          spark
+              .read()
+              .format(inputFileFormat)
+              .option("header", true)
+              .option("inferSchema", true)
+              .load(inputFileLocation);
+      inputData
+          .write()
+          .format("com.google.cloud.spark.bigquery")
+          .option("table", bqDataset + "." + bqTable)
+          .option("temporaryGcsBucket", bqTempBucket)
+          .mode(SaveMode.Append)
+          .save();
+    } catch (Exception th) {
+      LOGGER.error("Exception in GCStoBigquery", th);
+      JSONObject obj = new JSONObject();
+      obj.put("projectId", projectId);
+      obj.put("inputFileLocation", inputFileLocation);
+      obj.put("bqDataset", bqDataset);
+      obj.put("bqTable", bqTable);
+      obj.put("deadLetterQueue", dlq);
+      publishMessage(obj.toString());
+      if (Objects.nonNull(spark)) {
+        spark.stop();
+      }
+    }
```

Comment:
This is too broad, it captures every single exception in the full pipeline. I think you should capture the exceptions that you expect to handle, and let the rest of exceptions be handled (or not) by Spark. The job would fail anyways.

Reply:
For exception handling, The example spark is simple (toy example), hence for any exception it is Publish the error to Pub/Sub and stops. I am also writing errors from cloud function to Pub/Sub. Not by default, but cloud functions are tried multiple times(and hence the spark), so yes multiple and transient error might go be written in Pub/Sub. I will like the idea of making sure the DLQ having only single entry failed jobs, so pipelines can be retired with less human intervention. Please share if you have some examples/patterns, I will incorporate them. Meanwhile I will also make it clear in Readme.